sharedbuffers/mapped_struct.py

```diff
@@ -333,14 +335,185 @@ class mapped_dict(dict):
+    elif isinstance(key, (tuple, proxied_tuple, frozenset)):
+        hval = xxhash.xxh64(str(type(key))).intdigest()
+        for value in key:
+            hval = _mix_hash(hval, _stable_hash(value))
+    else:
+        raise TypeError("don't know how to hash key of type %s" % type(key))
+
+    # Truncating the hash value to 48 bits should have no practical impact
+    # on keys' destribution, and it helps us catch subtle overflow bugs
+    return hval & 0xffffffffffff
+
+
+def _enum_keys(obj):
+    idx = 0
+    for key in obj.iterkeys():
+        yield key, idx
+        idx += 1
+
+
+class BufferIO(object):
+    """
+    Gimmick object. Its whole purpose is to implement
+    the protocol expected by the IdMapper classes.
+    """
+
+    def __init__(self, buf, offs):
+        self.buf = buf
+        self.offs = offs
+        self.pos = 0
+
+    def write(self, value):
+        size = len(value)
+        offs = self.offs + self.pos
+        self.buf[offs:offs + size] = value
+        self.pos += size
+        return size
+
+    def tell(self):
+        return self.pos
+
+    def flush(self):
+        pass
+
+    def seek(self, pos):
+        self.pos = pos
+
+
+@cython.cclass
+class proxied_dict(object):
+
+    cython.declare(objmapper=object, vlist=proxied_list)
+
+    def __init__(self, objmapper, vlist):
+        self.objmapper = objmapper
+        self.vlist = vlist
+
+    @classmethod
+    def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
+        ipos = offs
+        offs += 8   # Reserve space for value-list offset.
+        iobuf = BufferIO(buf, offs)
+        mapper_size = ObjectIdMapper.build(_enum_keys(obj), iobuf, map_file=False)
+        offs += mapper_size
+        struct.pack_into("=Q", buf, ipos, mapper_size)
+        return offs + proxied_list.pack_into(obj.values(), buf, offs, idmap, implicit_offs)
+
+    @classmethod
+    def unpack_from(cls, buf, offs, idmap = None):
+        mapper_size, = struct.unpack_from("=Q", buf, offs)
+        offs += 8
+        objmapper = ObjectIdMapper.map_buffer(buf, offs)
+        offs += mapper_size
+        vlist = proxied_list.unpack_from(buf, offs, idmap)
+        return proxied_dict(objmapper, vlist)
+
+    def get(self, key, default_val = None):
+        idx = self.objmapper.get(key)
+        return self.vlist[idx] if idx is not None else default_val
+
+    def __getitem__(self, key):
+        idx = self.objmapper.get(key)
+        if idx is None:
+            raise KeyError(key)
+        return self.vlist[idx]
+
+    def __contains__(self, key):
+        return self.objmapper.get(key) is not None
+
+    def has_key(self, key):
+        return key in self
+
+    def iteritems(self):
+        for key, idx in self.objmapper.iteritems():
+            yield key, self.vlist[idx]
+
+    def items(self):
+        return list(self.iteritems())
+
+    def iterkeys(self):
+        return iter(self.objmapper)
+
+    def keys(self):
+        return list(self.iterkeys())
```

Comment:
You can just return `self.objmapper.keys()`

Also, check that you override the default `make_sequential=True` with False, I think it can be surprising if iterating a dict has to sort big arrays.